python/ql/src/semmle/python/security/dataflow/WeakSensitiveDataHashingCustomizations.qll

```diff
@@ -0,0 +1,157 @@
+    Cryptography::HashingAlgorithm algorithm;
+
+    WeakHashingOperationInputSink() {
+      exists(Cryptography::CryptographicOperation operation |
+        algorithm = operation.getAlgorithm() and
+        algorithm.isWeak() and
+        this = operation.getAnInput()
+      )
+    }
+
+    override string getAlgorithmName() { result = algorithm.getName() }
+  }
+}
+
+/**
+ * Provides default sources, sinks and sanitizers for detecting
+ * "use of a broken or weak cryptographic hashing algorithm on sensitive data"
+ * vulnerabilities on sensitive data that DOES require computationally expensive
+ * hashing, as well as extension points for adding your own.
+ *
+ * Also see the `NormalHashFunction` module.
+ */
+module ComputationallyExpensiveHashFunction {
+  /**
+   * A data flow source of sensitive data that requires computationally expensive
+   * hashing for "use of a broken or weak cryptographic hashing algorithm on sensitive
+   * data" vulnerabilities.
+   */
+  abstract class Source extends DataFlow::Node {
+    /** Gets the classification of the sensitive data. */
+    abstract string getClassification();
+  }
+
+  /**
+   * A data flow sink for sensitive data that requires computationally expensive
+   * hashing for "use of a broken or weak cryptographic hashing algorithm on sensitive
+   * data" vulnerabilities.
+   */
+  abstract class Sink extends DataFlow::Node {
+    /**
+     * Gets the name of the weak hashing algorithm.
+     */
+    abstract string getAlgorithmName();
+
+    /**
+     * Holds if this sink is for a computationally expensive hash function (meaning that
+     * hash function is just weak in some other regard.
+     */
+    abstract predicate isComputationallyExpensive();
+  }
+
+  /**
+   * A sanitizer of sensitive data that requires computationally expensive
+   * hashing for "use of a broken or weak cryptographic hashing
+   * algorithm on sensitive data" vulnerabilities.
+   */
+  abstract class Sanitizer extends DataFlow::Node { }
+
+  /**
+   * A source of passwords, considered as a flow source.
+   */
+  class PasswordSourceAsSource extends Source, SensitiveDataSource {
+    PasswordSourceAsSource() {
+      // TODO: once https://github.com/github/codeql/pull/5739 has been merged,
+      // don't use hardcoded value anymore
+      SensitiveDataSource.super.getClassification() = "password"
+    }
+
+    override string getClassification() { result = SensitiveDataSource.super.getClassification() }
+  }
+
+  /**
+   * The input to a password hashing operation using a weak algorithm, considered as a
+   * flow sink.
+   */
+  class WeakPasswordHashingOperationInputSink extends Sink {
+    Cryptography::CryptographicAlgorithm algorithm;
+
+    WeakPasswordHashingOperationInputSink() {
+      (
+        algorithm instanceof Cryptography::PasswordHashingAlgorithm and
+        algorithm.isWeak()
+        or
+        algorithm instanceof Cryptography::HashingAlgorithm
+      ) and
```

Comment:
Here I had to remind myself that a `PasswordHashingAlgorithm` cannot be a `HashingAlgorithm`. It might be nice to have the disjointness in the names, but I can only come up with silly examples such as `ExpensiveHashingAlgorithm`/`CheapHashingAlgorithm`...

Reply:
I agree that this is _a bit_ surprising. We should not try to change this naming in this PR, although we could try to do that in an other PR. 

This modeling and therefore naming is now shared with JS (yay :tada:), so you would have to come up with a really good name for this to be worth it I think :smile:

Reply:
Perhaps a simple comment for now, then..

Reply:
sure :+1: (done)